src/BaseControllerV2.ts

```diff
@@ -87,3 +101,31 @@ export class BaseController<S extends Record<string, any>> {
     this.internalListeners.clear();
   }
 }
+
+// This function acts as a type guard. Using a `typeof` conditional didn't seem to work.
+function isAnonymizingFunction<T>(x: boolean | Anonymizer<T>): x is Anonymizer<T> {
+  return typeof x === 'function';
+}
+
+export function getAnonymizedState<S extends Record<string, any>>(state: S, schema: Schema<S>) {
```

Comment:
It'd be better for `S` to extend `Record<string, unknown>` as opposed to `Record<string, any>`. See [here](https://github.com/Microsoft/TypeScript/pull/24439) for details on `unknown`.

This is a change I believe that we should make everywhere in `BaseControllerV2`, and one we can do in a follow-up.

Reply:
In addition, both `getAnonymizedState` and `getPersistentState` should probably take an additional generic type parameter, which should be returned instead of `Partial<S>`. If we always return `Partial<S>`, consumers will have to check for the existence of every key of `S` if they try to work with the return value of these functions.

Reply:
Definitely in favour of using `unknown` over `any`!

I'm not sure about adding an additional generic type to those functions though :thinking:. The main idea with the persisted and anonymous state is that we _wouldn't_ be working with the return values. The persisted state gets persisted, and the anonymized state gets sent off as a state blob.